File: public/books/fraud-detection/code/01_data_overview.R
# =====================================================
# code/01_data_overview.R
# 第 1 章：数据概览与零模型基线
# 输入：data/bao2020_full.csv
# 输出：控制台数字（写入 _NUMBERS.md Chap 1 节）
# =====================================================

library(tidyverse)
library(here)
set.seed(2026)

d <- read_csv(here::here("data", "bao2020_full.csv"),
              show_col_types = FALSE)

# 全样本规模
cat(sprintf("全样本: %d firm-years, %d 公司, fyear %d-%d\n",
            nrow(d), n_distinct(d$gvkey),
            min(d$fyear), max(d$fyear)))

cat(sprintf("misstate=1: %d (%.4f%%)\n",
            sum(d$misstate), 100 * mean(d$misstate)))

# Bao 时间切分
splits <- list(
  train = d %>% filter(fyear >= 1991, fyear <= 2002),
  valid = d %>% filter(fyear >= 2003, fyear <= 2008),
  test  = d %>% filter(fyear >= 2009, fyear <= 2014)
)
for (name in names(splits)) {
  s <- splits[[name]]
  cat(sprintf("%s: n=%d, fraud=%d, rate=%.3f%%\n",
              name, nrow(s), sum(s$misstate),
              100 * mean(s$misstate)))
}

# Enron (gvkey=6127) 关键年度
cat("\n--- Enron (gvkey=6127) ---\n")
d %>% filter(gvkey == 6127) %>%
  arrange(fyear) %>%
  select(fyear, misstate, p_aaer, at, sale, ni) %>%
  print(n = Inf)

# 零模型在测试集上的指标
test <- splits$test
n_test <- nrow(test)
n_pos  <- sum(test$misstate)
cat(sprintf("\n零模型: AUC=0.500, Recall@1%%=0, Precision@1%%=0\n"))
cat(sprintf("测试集 n=%d, 阳性=%d, 1%% 名额=%d\n",
            n_test, n_pos, ceiling(n_test * 0.01)))

File: public/books/fraud-detection/code/01_data_overview.py
"""
code/01_data_overview.py
第 1 章：数据概览与零模型基线（Python 等价实现）
"""
import random
import numpy as np
import pandas as pd

random.seed(2026)
np.random.seed(2026)

d = pd.read_csv("data/bao2020_full.csv")

print(f"全样本: {len(d):,} firm-years, "
      f"{d['gvkey'].nunique():,} 公司, "
      f"fyear {int(d['fyear'].min())}-{int(d['fyear'].max())}")
print(f"misstate=1: {int(d['misstate'].sum())} "
      f"({100 * d['misstate'].mean():.4f}%)")

splits = {
    "train": d.query("1991 <= fyear <= 2002"),
    "valid": d.query("2003 <= fyear <= 2008"),
    "test":  d.query("2009 <= fyear <= 2014"),
}
for name, s in splits.items():
    print(f"{name}: n={len(s):,}, fraud={int(s['misstate'].sum())}, "
          f"rate={100 * s['misstate'].mean():.3f}%")

print("\n--- Enron (gvkey=6127) ---")
print(d.query("gvkey == 6127")
      .sort_values("fyear")[["fyear", "misstate", "p_aaer",
                              "at", "sale", "ni"]]
      .to_string(index=False))

test = splits["test"]
print(f"\n零模型: AUC=0.500, Recall@1%=0, Precision@1%=0")
print(f"测试集 n={len(test):,}, 阳性={int(test['misstate'].sum())}, "
      f"1% 名额={int(np.ceil(len(test) * 0.01))}")
